### PR TITLE
[FLINK-18378] Improve CatalogTable schema resolution

### PR DESCRIPTION
**Pull request by dawidwys**
# What is the purpose of the change

Make it possible to use different `CatalogTable` implementation and properly resolve such tables. Moreover apply the schema fields resolution for fields of views.

## Brief change log

  - Pass the resolved schema as part of the `TableLookupResult`
  - When resolving proctime time attribute use the nullability of the resolved expression instead of the original field (otherwise if user declares proctime field nullable, there will be mismatch)

## Verifying this change

Added tests in `JavaCatalogTableITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by dawidwys** [comment]
Could you take a look @wuchong @godfreyhe

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableITCase.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+import java.util.Optional;
+
+import static org.apache.flink.table.api.Expressions.$;
+import static org.apache.flink.table.api.Expressions.lit;
+
+/**
+ * Tests for resolving types of computed columns (including time attributes) of tables from catalog.
+ */
+@RunWith(Parameterized.class)
+public class JavaCatalogTableITCase extends TableTestBase {
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableITCase.java:53]: Sure I will rename the test. 

Nevertheless it is still an IT test imo. It does not test a single module, but an integration of few (catalogmanager, parser, planner, ....)

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/CatalogTableSchemaResolver.java [hunk 2]
```diff
@@ -76,16 +78,16 @@ public TableSchema resolve(TableSchema tableSchema) {
...
 			DataType fieldType = fieldTypes[i];
-			if (tableColumn.isGenerated() && isProctimeType(tableColumn.getExpr().get(), tableSchema)) {
+			DataType exprType = null;
+			if (tableColumn.isGenerated()) {
+				exprType = resolveExpressionDataType(tableColumn.getExpr().get(), tableSchema);
+			}
+
+			if (exprType != null && isProctime(exprType)) {
 				if (fieldNames[i].equals(rowtime)) {
 					throw new TableException("Watermark can not be defined for a processing time attribute column.");
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/CatalogTableSchemaResolver.java:88]: We need the check for `isProctime(...)`.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/CatalogTableSchemaResolver.java [hunk 3]
```diff
@@ -76,16 +78,16 @@ public TableSchema resolve(TableSchema tableSchema) {
...
 				}
-				TimestampType originalType = (TimestampType) fieldType.getLogicalType();
-				LogicalType proctimeType = new TimestampType(
-						originalType.isNullable(),
-						TimestampKind.PROCTIME,
-						originalType.getPrecision());
-				fieldType = TypeConversions.fromLogicalToDataType(proctimeType);
+				fieldType = exprType;
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/CatalogTableSchemaResolver.java:90]: I was also thinking about it and I can do that. Nevertheless in my opinion this requires careful consideration.

What is the point of data type of a computed column, if we completely ignore it?

[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/CatalogTableSchemaResolver.java]: The case is that the return type of functions (built-in or UDF) may change in the future, e.g. we are planning to change the return type of `CURRENT_TIMESTAMP` to `TIMESTAMP WITH LCOAL TIME ZONE`.

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/CatalogTableSchemaResolver.java]: I understand the goal and as I said I will change it the way you asked. Just wanted to point out that I think we do not have a clear semantic of the type of a computed column.

[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/CatalogTableSchemaResolver.java]: Yes.. We didn't have a public discussion about this and some people may thought the type is snapshotted when CREATE TABLE. Maybe we should write this down in documentation.

**Pull request closed without merging**